include/execution.hpp

```diff
@@ -1873,6 +1860,251 @@ namespace std::execution {
+              using WhenAll = __member_t<CvrefReceiverId, __sender>;
+              using Receiver = __t<decay_t<CvrefReceiverId>>;
+              template <size_t Index>
+                struct __receiver : receiver_adaptor<__receiver<Index>> {
+                  Receiver&& base() && noexcept {
+                    return (Receiver&&) __op_->__recvr_;
+                  }
+                  const Receiver& base() const & noexcept {
+                    return __op_->__recvr_;
+                  }
+                  template <class... Values>
+                    void set_value(Values&&... vals) && noexcept try {
+                      __state_t expected = __started;
+                      // We only need to bother recording the completion values
+                      // if we're not already in the "error" or "done" state.
+                      if (__op_->__state_ == __started) {
+                        std::get<Index>(__op_->__values_).emplace(
+                            (Values&&) vals...);
+                      }
+                      (void) __op_->__barrier_.arrive();
+                    } catch(...) {
+                      ((__receiver&&) *this).set_error(current_exception());
+                    }
+                  template <class Error>
+                      requires receiver<Receiver, Error>
+                    void set_error(Error&& error, __state_t expected = __started) && noexcept {
+                      // TODO: What memory orderings are actually needed here?
+                      if (__op_->__state_.compare_exchange_strong(expected, __error)) {
+                        // We won the race, free to write the error into the operation
+                        // state without worry.
+                        try {
+                          __op_->__errors_.template emplace<decay_t<Error>>((Error&&) error);
+                        } catch(...) {
+                          __op_->__errors_.template emplace<exception_ptr>(current_exception());
+                        }
+                        __op_->__stop_source_.request_stop();
+                      }
+                      (void) __op_->__barrier_.arrive(); // Fatal error if it throws
+                    }
+                  void set_done() && noexcept try {
+                    __state_t expected = __started;
+                    // Transition to the "done" state if and only if we're in the
+                    // "started" state. (If this fails, it's because we're in an
+                    // error state, which trumps cancellation.)
+                    if (__op_->__state_.compare_exchange_strong(expected, __done)) {
+                      __op_->__stop_source_.request_stop();
+                    }
+                    (void) __op_->__barrier_.arrive(); // potentially throwing
+                  } catch(...) {
+                    // Try to put this operation into the error state
+                    ((__receiver&&) *this).set_error(current_exception(), __done);
```

Comment:
I think that if 'done' won the race, then errors are irrelevant and should be discarded. the request to discard results 'happens-before' the error

Reply:
Right, but we only get here if trying to transition to the "done" state has failed. Maybe when `std::barrier::arrive` throws, I should just consider the error fatal?

Reply:
if the compare exchange fails then the error would own the call to arrive(), so done should not call arrive.

If done() wins the compare exchange race then it should own the call to arrive.

if arrive throws, then attempting to shift to an error state will only call arrive again. might as well loop on arrive in done.. I think terminate is a better option than trying to move from the done state to the error state.

Reply:
I'll be replacing the `barrier` with a simple atomic ref count. That failure mode will simply go away, and this code gets simpler.